### ISSUE TITLE
map@R metric implem seems mistaken

### ISSUE THREAD

**Issue by drasros** [MalongTech/research-xbm#11]
Hi! 
Thanks for releasing this repo!
I wanted to let you know that I am pretty sure there is a mistake in the implementation of mAP@R:
[Line 35](https://github.com/msight-tech/research-xbm/blob/223ecdc25f71ef1721a58bc87cc567025a32bc92/ret_benchmark/data/evaluations/eval.py#L35), you divide by max_possible_matches_per_row for this query (which is R). 
But what should be done is divide by the _actual_ number of relevant items in the R-best-ranked ones. 
see [wikipedia page on AP](https://en.wikipedia.org/w/index.php?title=Information_retrieval&oldid=793358396#Average_precision):
![image](https://user-images.githubusercontent.com/16518885/119517332-d0657500-bd77-11eb-8d97-2ed40ac2558b.png)
I also wrote some tests against the [sklearn implem](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.average_precision_score.html) of AP (I can send them if needed). 
This mistake seems to also be present in the repo of 'Deep metric learning reality check' with very similar code, I'm not sure who was inspired from who, but I will open an issue there too. 
Regards, 
A

**Comment by drasros** [comment]
After discussion with Kevin Musgrave, there is a confusion with the definition of AP@R which can be understood in 2 ways: 
* truncated-sum full-AP
* AP over subset of top-R

As a result, this implem disagrees an implem using sklearn AP over top-R. 
I think both versions are fine, it is just good to know they are different. 
[https://github.com/KevinMusgrave/pytorch-metric-learning/issues/331](https://github.com/KevinMusgrave/pytorch-metric-learning/issues/331)
